File: backend/app/services/provider_matching.py
"""Match appointments to a provider record (handles legacy General Practice bookings)."""

from __future__ import annotations

from typing import Any

_UNASSIGNED_LABELS = frozenset({"", "General Practice", "Provider"})


def _base_specialty(specialty: str) -> str:
    return specialty.split("·")[0].strip()


def appointment_matches_provider(appt: dict[str, Any], provider: dict[str, Any]) -> bool:
    pname = str(appt.get("provider_name") or "General Practice").strip()
    display = str(provider.get("display_name") or "").strip()
    specialty = str(provider.get("specialty") or "General Practice").strip()
    base = _base_specialty(specialty)

    if pname == display:
        return True
    if pname in {specialty, base}:
        return True
    if pname in _UNASSIGNED_LABELS:
        return True
    return False

File: frontend/src/lib/scheduling/providerMatching.ts
const UNASSIGNED_LABELS = new Set(["", "General Practice", "Provider"]);

function baseSpecialty(specialty: string): string {
  return specialty.split("·")[0]?.trim() ?? specialty;
}

export interface ProviderMatchFields {
  display_name: string;
  specialty?: string | null;
}

export interface AppointmentProviderFields {
  provider_name?: string | null;
}

/** True when an appointment belongs to this doctor (includes legacy unassigned bookings). */
export function appointmentMatchesProvider(
  appointment: AppointmentProviderFields,
  provider: ProviderMatchFields | undefined,
): boolean {
  if (!provider?.display_name) return false;

  const pname = (appointment.provider_name ?? "General Practice").trim();
  const display = provider.display_name.trim();
  const specialty = (provider.specialty ?? "General Practice").trim();
  const base = baseSpecialty(specialty);

  if (pname === display) return true;
  if (pname === specialty || pname === base) return true;
  if (UNASSIGNED_LABELS.has(pname)) return true;
  return false;
}
